fendermint/vm/interpreter/src/genesis.rs

```diff
@@ -0,0 +1,795 @@
+use std::path::{Path, PathBuf};
+use std::pin::Pin;
+use std::sync::Arc;
+
+use anyhow::{anyhow, Context};
+use cid::Cid;
+use ethers::abi::Tokenize;
+use ethers::core::types as et;
+use fendermint_actor_eam::PermissionModeParams;
+use fendermint_eth_hardhat::{Hardhat, FQN};
+use fendermint_vm_actor_interface::diamond::{EthContract, EthContractMap};
+use fendermint_vm_actor_interface::eam::EthAddress;
+use fendermint_vm_actor_interface::ipc::IPC_CONTRACTS;
+use fendermint_vm_actor_interface::{
+    account, burntfunds, chainmetadata, cron, eam, init, ipc, reward, system, EMPTY_ARR,
+};
+use fendermint_vm_core::{chainid, Timestamp};
+use fendermint_vm_genesis::{ActorMeta, Genesis, Power, PowerScale, Validator};
+use futures_util::io::Cursor;
+use fvm::engine::MultiEngine;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_car::{load_car, CarHeader};
+use fvm_ipld_encoding::CborStore;
+use fvm_shared::chainid::ChainID;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::version::NetworkVersion;
+use ipc_actors_abis::i_diamond::FacetCut;
+use num_traits::Zero;
+
+use crate::fvm::state::snapshot::{derive_cid, StateTreeStreamer};
+use crate::fvm::state::{FvmGenesisState, FvmStateParams};
+use crate::fvm::store::memory::MemoryBlockstore;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use tokio_stream::StreamExt;
+use tokio_util::compat::TokioAsyncWriteCompatExt;
+
+/// The sealed genesis state metadata
+#[serde_as]
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
+struct GenesisMetadata {
+    pub state_params: FvmStateParams,
+    pub validators: Vec<Validator<Power>>,
+}
+
+impl GenesisMetadata {
+    fn new(state_root: Cid, out: GenesisOutput) -> GenesisMetadata {
+        let state_params = FvmStateParams {
+            state_root,
+            timestamp: out.timestamp,
+            network_version: out.network_version,
+            base_fee: out.base_fee,
+            circ_supply: out.circ_supply,
+            chain_id: out.chain_id.into(),
+            power_scale: out.power_scale,
+            app_version: 0,
+        };
+
+        GenesisMetadata {
+            state_params,
+            validators: out.validators,
+        }
+    }
+}
+
+pub async fn read_genesis_car<DB: Blockstore + 'static + Send + Sync>(
+    bytes: Vec<u8>,
+    store: &DB,
+) -> anyhow::Result<(Vec<Validator<Power>>, FvmStateParams)> {
+    let roots = load_car(store, Cursor::new(&bytes)).await?;
+
+    if roots.len() != 1 {
+        return Err(anyhow!("invalid genesis car, should have 1 root cid"));
+    }
+
+    let metadata_cid = roots[0];
+    let metadata = if let Some(metadata) = store.get_cbor::<GenesisMetadata>(&metadata_cid)? {
+        metadata
+    } else {
+        return Err(anyhow!("invalid genesis car, metadata not found"));
+    };
+
+    Ok((metadata.validators, metadata.state_params))
+}
+
+/// The output of genesis creation
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct GenesisOutput {
```

Comment:
It seems like there is exactly same struct in in fvm package called `FvmGenesisOutput`. Maybe you can re-use?

Reply:
I think that one is in a different mod. I'm planning to remove the old implementation after this one in favour of current implementation.